### ISSUE TITLE
AADSTS90009 (I think) due to no scope form parameter on refresh

### ISSUE THREAD

**Issue by nedgrady** [AzureAD/microsoft-authentication-library-for-js#5198]
### Core Library

MSAL.js v2 (@azure/msal-browser)

### Core Library Version

2.28.1

### Wrapper Library

MSAL React (@azure/msal-react)

### Wrapper Library Version

1.4.0

### Public or Confidential Client?

Public

### Description

When the library makes a refresh request (using `aquireTokenSilent`) I get a 400 error back from the _/token_ endpoint. Initially obtaining the token works fine with these code samples, and when I refresh local storage I can get a token again.

When the library makes a refresh request (e.g. when I invoke `getServiceToken` [see code samples] after initial token has expired) I can see following network request in dev tools:

> POST https://login.microsoftonline.com/bbbbbbbb-663d-4ae5-9253-cb7d2aac234e/oauth2/v2.0/token
> Payload (form data):
> client_id: aaaaaaaa-663d-4ae5-9253-cb7d2aac234e
> scope: openid profile offline_access
> grant_type: refresh_token
> client_info: 1
> refresh_token: 0.ABC...

Which gives the following response
400 
`{
     "error" :  "invalid_request",
     "error_description":  "AADSTS90009: Application 'aaaaaaaa-663d-4ae5-9253-cb7d2aac234e' (aaaaaaaa-663d-4ae5-9253-cb7d2aac234e) is requesting a token for itself. This scenario is supported only if resource is specified using the GUID based App Identifier. ..." ...
}`

Now if I make the same request using postman, but replace the scope form parameter with
> scope: https://back-end-service.com/FRONT-END

I get a 200 OK response containing a token & new refresh token as expected.

When `getServiceToken` is invoked, I can not only see the POST returning a 400 error, but there is an uncaught console error. If I add an ErrorBoundary type component, it catches the error.. but the try.. catch inside `getServiceToken` never catches an error.

I suppose my question are is it possible to
* Stop this 400 error from happening by passing the 'correct' scope to the resfresh_token request
* Supressing the error somehow?
 Or is the error a symptom and not a root cause?

N.b. the initial POST looks like this:
POST https://login.microsoftonline.com/bbbbbbbb-663d-4ae5-9253-cb7d2aac234e/oauth/v2.0/token?resource=aaaaaaaa-663d-4ae5-9253-cb7d2aac234e&scope=https://back-end-service.com/FRONT-END

Which returns 200OK and a token.

### MSAL Configuration

```javascript
// n.b. this is hand-typed as I can't copy + paste off corporate network
{
    auth: {
        // Not the real clientId
        clientId: "aaaaaaaa-663d-4ae5-9253-cb7d2aac234e",
        // not real tenantId appended to the end
        authority: "https://login.microsoftonline.com/bbbbbbbb-663d-4ae5-9253-cb7d2aac234e",
        redirectUri: window.location.origin
    },
    cache: "localStorage"

}
```

### Relevant Code Snippets

```javascript
// msalConfig.js:

// exported for use in react app in MsalProvider
// _.once ensures the function only runs once
export const getMsalInstance = _.once(() => {
    const msalInstance = new PublicClientApplication(msalConfig) //see MSAL Configuration section
    
    msalIntance.addEventCallback(event => {
        if(event.eventType === EventType.LOGIN_SUCCESS)
            msalInstance.setActiveACcount(event.payload.account)
    })
})

// n.b. httpRequestService.js is used OUTSIDE of hooks, so never uses @azure/msal-react
export async function getServiceToken() {
    const request = {
        scopes: ["https://back-end-service.com/FRONT-END"],
        account: getMsalInstance().getActiveAccount()
    }

    try {
        console.log(reqeust) //can always see initialized account
        const result = await getMsalInstance().aquireTokenSilent(request)
        console.log(result) // can always see a result, even with a new token (e.g. exp is set in the future here)
        return result.accessToken
    } catch (err) {
        // this catch handler never gets hit!!!
    }
}

// httpRequestService.js:
export async function(url) {
    let authHeader = "Bearer " + (await getServiceToken())
    console.log(authHeader) // I can even see the authHeader here with the new token!!
    return axios.get(url, { headers: { Authorization: authHeader } })
}
```

### Identity Provider

Azure AD / MSA

### Source

External (Customer)

**Comment by sameerag** [comment]
@nedgrady The error indicates improper scopes for the token requested and it is thrown by eSTS. MSAL JS is only communicating the error to you.

> Stop this 400 error from happening by passing the 'correct' scope to the resfresh_token request
> Supressing the error somehow?
> Or is the error a symptom and not a root cause?

 `getServiceToken` calls the `acquireTokenSilent` which basically tries to renew your session if the `refresh_token` is available and the `access_token` in the cache is expired. When that fails, it throws.  However, I see that there could be a case where the error is just emitted to the console and not bubbled up, hence suppressing it.

Please watch out this space for an update. Will raise a PR once I discuss this with the team.

**Comment by nedgrady** [comment]
Thanks for the response @sameerag. Really appreciate the thoughts & help.

> @nedgrady The error indicates improper scopes for the token requested and it is thrown by eSTS. MSAL JS is only communicating the error to you.

Is there a way I can ensure the correct scope is supplied when the session is renewed? (e.g. the same scope as when the token is initially retrieved)

**Comment by sameerag** [comment]
@nedgrady 
On second look, I think your code is behaving as expected. I missed the comment you added that the silent call always works.
>         console.log(result) // can always see a result, even with a new token (e.g. exp is set in the future here)

Does this mean that the token acquisition is indeed successful?

There are some cases where if the fetch fails from `/token` endpoint, `msal js` retries the request with `/authorize`. I would like to check the fiddler trace for your use case to actually see what is happening as the specific error you shared above looks new.

My email is in my profile to share the fiddler trace.

**Comment by nedgrady** [comment]
Have shared some logs with you.

> 
> > ```
> >     console.log(result) // can always see a result, even with a new token (e.g. exp is set in the future here)
> > ```
> 
> Does this mean that the token acquisition is indeed successful?
> 

Yeah I can see a token returned by getServiceToken(), but still see an uncaught error in the console. If I wrap an error boundary around my whole app:

```
class ErrorBoundary extends React.Component {
    constructor (props) {
        super (props); this.state = { hasError: false };
     }
    
    static getDerivedStateFromError() {  return { hasError: true }; }

    componentDidCatch(error, errorInfo) {
        console.log("error caught, dumping getServiceToken() start: ")
        getServiceToken().then(token => {
            console.log(token);
            console.log("error caught, dumping getServiceToken() end: ");
        });
    }
    render() {
        if(this.state.hasError) return <h1>Something went wrong.</h1>
     
        return this.props.children;
    }
}

ReactDOM.render(
    <ErrorBoundary>
        <MsalProvider instance={getMsalInstance()}>
            <MsalAuthenticationTemplate>
                 <AuthenticatedTemplate>
                    <>
                        <p>only see this before the token expires for the first time</p>
                         <App />
                    </>
                 </AuthenticatedTemplate>
                  <UnauthenticatedTemplate>
                      <p>never see this</p>
                 </UnauthenticatedTemplate>
            <MsalAuthenticationTemplate interactionType={InteractionType.Redirect}>
        </MsalProvider>
    </ErrorBoundary>
)
```

What I see in the console is:

> error caught, dumping getServiceToken() start:
> eyJ0ex....
> error caught, dumping getServiceToken() end: 

And I see the "Something went wrong." error on screen.

Very confusing.

**Comment by hada79** [comment]
I am having a similar experience: msal-browers 2.21.0, msal-react 1.4.5
I have an api and SPA sharing the same app registration. Like Ned, I can get the initial token fine, when trying to refresh the token, I get the 90009 error. The scopes I am sending are openid profile offline_access. 

Looking at [this post](https://stackoverflow.com/questions/42006468/oauth-2-0-and-azure-active-directory-error-aadsts90009), it seems like in order to share the app registration, I need to specify the client_id as the resource, however there doesn't appear to be a resource parameter in the config passed to the PCA? @nedgrady are you by chance sharing an app registration between a front end and backend?

**Comment by nedgrady** [comment]
> @nedgrady are you by chance sharing an app registration between a front end and backend?

Yeah @hada79 that's exactly what we are doing. Ithink being able to use the client_id as a resource would fix the problem.

**Comment by sameerag** [comment]
@hada79 have you been able to solve your problem by using `client_id` as a resource as @nedgrady suggested? 

@nedgrady Apologies for the delay, can you resend the logs to my email? I could not find them and not sure if I ever received them.

**Comment by hada79** [comment]
@sameerag There doesnt appear to be a way to specify the client ID as the resource using msal, which is the problem.

**Comment by nedgrady** [comment]
> @nedgrady Apologies for the delay, can you resend the logs to my email? I could not find them and not sure if I ever received them.

No probs no rush, not like our current auth solution is going to stop working. Have resent the email from my gmail account

**Comment by hada79** [comment]
@sameerag Any ideas on how to specify a resource id using msal? I believe you may have misread the thread, thinking nedgrady had provided a solution.

**Comment by dzendras** [comment]
I seem to be experiencing the same problem.

**Comment by sameerag** [comment]
Let me summarize the issue:
* App registration is shared between front-end and back-end
* An internal MSAL JS error is seen in these use cases in the browser console but the token acquisition succeeds eventually.

Is this description accurate @hada79?

**Comment by hada79** [comment]
@sameerag The refresh token acquisition never succeeds.

**Comment by obaileymwks** [comment]
I have encountered this issue as well. Have there been any developments on this not associated to this issue?

**Comment by sameerag** [comment]
@obaileymwks Is it on the latest `msal-browser`?

**Comment by obaileymwks** [comment]
@sameerag We're on v2.31.0
![image](https://github.com/AzureAD/microsoft-authentication-library-for-js/assets/104217912/508f068f-b28e-4104-9e30-aa6a910fb7e8)

**Comment by chrisincolorado** [comment]
I'm also seeing this issue on v2.31.0 where we have an app registration shared between front and back end.

**Comment by obaileymwks** [comment]
After not seeing any movement on this item for so long I went experimenting and found that these issues were resolved by creating an app registration for the front end to run as that has permissions to the app registration for your back end. This will cause you some problems if you're using a single app service in order to use App Roles in both front end and back end. Having two app services did make it so that we are no longer seeing `AADSTS90009` in my testing though.

**Comment by chrisincolorado** [comment]
Got it, good to know having separate registrations does fix it. We do have shared App Roles though, so hoping to avoid that.

**Comment by chrisincolorado** [comment]
@sameerag any chance this issue could get addressed? Managing App Roles for our backend and frontend separately would significantly complicate our role management.

**Comment by derisen** [comment]
@chrisincolorado your frontend and backend apps can share the same app registration, which would allow you to avoid duplicating app role definitions. See for instance: [Angular SPA calling a protected web API using App Roles to implement RBAC](https://github.com/Azure-Samples/ms-identity-javascript-angular-tutorial/tree/main/5-AccessControl/1-call-api-roles)

**Comment by chrisincolorado** [comment]
@derisen this is exactly what we're doing, except using MSAL React. It results in the `requesting a token for itself` errors described above.

**Comment by derisen** [comment]
@chrisincolorado hmm -to confirm, did you try exposing a scope and acquiring a token for that scope, as mentioned here in the registration steps: https://github.com/Azure-Samples/ms-identity-javascript-angular-tutorial/tree/main/5-AccessControl/1-call-api-roles#register-the-app-msal-angular-app

An implementation with msal-react can be found [here]( https://github.com/Azure-Samples/ms-identity-javascript-react-tutorial/tree/main/5-AccessControl/1-call-api-roles). I can confirm this to be working as expected

**Comment by tnorling** [comment]
@chrisincolorado Just checking in if the sample @derisen linked helped resolve this for you?

**Comment by chrisincolorado** [comment]
Hi @tnorling, I've checked that the configuration in the code matches these examples and we're still seeing that error. I still need to check if perhaps there's something configured differently in the Azure portal.

**Comment by ChilloManiac** [comment]
I'm also still seeing this issue, can this issue be reopened?

**Comment by YehonatanKadosh** [comment]
Same here!

**Comment by anwtn** [comment]
I encountered an issue despite not using MSAL. My app registration provided settings for user sign-up and was also employed by the backend API to call the Microsoft token endpoint for the authorization code flow.

Everything worked smoothly until I integrated custom attributes via the Enterprise App SSO configuration, which necessitates a verified domain. I configured this in the publisher settings and the Application ID URL under the "App Registration > Expose an API" blade. The latter's tooltip states:

> The Application ID URI, or identifier URI, is a globally unique URI used to identify the web API. This URI is the prefix for scopes in the OAuth protocol. You can use the default value `api://<application-client-id>` or specify a more readable URI.

Default App ID URLs look like: `api://<application-client-id>`. However, I configured mine as `https://my-entra-verified-domain.here`. After this custom domain configuration, I encountered an error when requesting an access token:

> <application-client-id> is requesting a token for itself. This scenario is supported only if the resource is specified using the GUID-based App Identifier.

The insight: This previously worked with the App ID URL as the application-client ID but failed after changing it. The conclusion: Changing to a custom App ID URL likely impacted some permission.

Here's what resolved it:

1. Add a scope under "App Registration" > "Expose an API" (e.g., named `this` for a resulting scope of `https://my-entra-verified-domain.here/this`).
2. At the bottom, add a client application using <application-client-id>.
3. Set the scope to `this` and perform Admin consent.
4. Under App Registration, go to the API Permissions blade.
5. Add a permission and select the desired API (<application-client-id>).
6. Click Grant Admin Consent.

In summary:
- The application can authenticate against itself by default.
- Changing the App ID URL requires explicit permission for the app to access itself using a new scope.

Note: After these changes, I successfully used `https://my-entra-verified-domain.here/this` and the default `https://my-entra-verified-domain.here/.default` scope. Replace `app://<application-client-id>` with your custom scope when in the authorization URL the user is sent to when requesting the authorization code.
